### PR TITLE
Dual headed fork choice [Reloaded]

### PR DESCRIPTION
**Pull request by mratsim**
Reboot of #1163 after we had to revert in #1181

The issue has been reproduced and transformed into a test case:

1. We sync
2. We go pass a finalization point, and so prune the fork choice
3. We receive an old block from before the finalization point
4. The fork choice crashes because the block's parent doesn't exist anymore in its history

The solution is like "Old" blocks, to also drops duplicates here:
https://github.com/status-im/nim-beacon-chain/blob/206ff7f9d6085d9e192644513626feed9d28f3d8/beacon_chain/block_pools/clearance.nim#L110-L142
https://github.com/status-im/nim-beacon-chain/blob/206ff7f9d6085d9e192644513626feed9d28f3d8/beacon_chain/beacon_node.nim#L550-L576

The stacktrace of the test after 9af7e06 is
image

which is the same as the one reported in https://github.com/status-im/nim-beacon-chain/pull/1163#issuecomment-644531631

```
_slot=143436 request_step=1 wall_clock_slot=145504
DBG 2020-06-16 06:54:41+02:00 Requesting blocks from peer                topics="syncman" tid=1901944 file=sync_manager.nim:611 peer=16Uiu2HAkucMAVckq8yyLU51pbnzjTZ6JptPLKdpH5USBGuuvinHP peer_score=200 slot=143436 slot_count=4 step=1
DBG 2020-06-16 06:54:41+02:00 Synchronization loop waiting for new peer  topics="syncman" tid=1901944 file=sync_manager.nim:935 local_head_slot=143432 sleep_time=6m24s0ns wall_head_slot=145504 workers_count=2
DBG 2020-06-16 06:54:41+02:00 Received blocks on request                 topics="syncman" tid=1901944 file=sync_manager.nim:762 blocks_count=4 blocks_map=xxxx peer=16Uiu2HAkvfDTKtFoSU5CHoA47BRwcJhN2zfvmr1kZXAwbweV3pq2 peer_score=200 request_count=4 request_slot=143432 request_step=1
DBG 2020-06-16 06:54:41+02:00 Forward sync imported blocks               topics="beacnde" tid=1901944 file=beacon_node.nim:552 count=4 local_head_slot=143432
DBG 2020-06-16 06:54:41+02:00 Block received                             topics="beacnde" tid=1901944 file=beacon_node.nim:293 blockRoot=1beab7a4 cat=block_listener pcs=receive_block signedBlock="(slot: 143432, proposer_index: 734, parent_root: \"c276ba51\", state_root: \"9c33444b\", proposer_slashings_len: 0, attester_slashings_len: 0, attestations_len: 7, deposits_len: 0, voluntary_exits_len: 0)"
DBG 2020-06-16 06:54:41+02:00 Block already exists                       topics="clearance" tid=1901944 file=clearance.nim:133 blck="(slot: 143432, proposer_index: 734, parent_root: \"c276ba51\", state_root: \"9c33444b\", proposer_slashings_len: 0, attester_slashings_len: 0, attestations_len: 7, deposits_len: 0, voluntary_exits_len: 0)" blockRoot=1beab7a4 cat=filtering pcs=block_addition
 peers: 0 ❯ finalized: abfa25a6:4034 ❯ head: 1beab7a4:4482:8 ❯ time: 4546:1 (145473)                                                                                                                                                                           ETH: 31.47697782 Traceback (most recent call last, using override)
../csu/libc-start.c(308) 
/home/arnetheduck/status/nim-beacon-chain/beacon_chain/mainchain_monitor.nim(526) main
/home/arnetheduck/status/nim-beacon-chain/beacon_chain/mainchain_monitor.nim(519) NimMain
/home/arnetheduck/status/nim-beacon-chain/beacon_chain/beacon_node.nim(1109) main
/home/arnetheduck/status/nim-beacon-chain/beacon_chain/beacon_node.nim(851) start
/home/arnetheduck/status/nim-beacon-chain/vendor/nim-chronicles/chronicles.nim(331) run
/home/arnetheduck/status/nim-beacon-chain/vendor/nimbus-build-system/vendor/Nim/lib/system/excpt.nim(407) reportUnhandledError
/home/arnetheduck/status/nim-beacon-chain/vendor/nimbus-build-system/vendor/Nim/lib/system/excpt.nim(358) reportUnhandledErrorAux
Error: unhandled exception: Trying to acces value with err Result: process_block_error: (kind: fcErrUnknownParent, child_root: 1BEAB7A4C29A864B05F3D8E5AAC8727F34D06D28280BD89D5BDE969B436D54DE, parent_root: C276BA5155F19C72D73932FE75D8291DD4FC3630A5FDF82D3A3484319BFA04CD) [ResultDefect]
```

### CODE REVIEW

File: beacon_chain/attestation_pool.nim [hunk 1]
```diff
@@ -22,10 +23,23 @@ func init*(T: type AttestationPool, blockPool: BlockPool): T =
   # TODO blockPool is only used when resolving orphaned attestations - it should
   #      probably be removed as a dependency of AttestationPool (or some other
   #      smart refactoring)
+
+  # TODO: In tests, on blockpool.init the finalized root
+  #       from the `headState` and `justifiedState` is zero
+  let forkChoice = initForkChoice(
+    finalized_block_slot = default(Slot),             # This is unnecessary for fork choice but may help external components
```
Comment (arnetheduck) [inline beacon_chain/attestation_pool.nim:30]: btw, I never understood this comment - which external components are helped here?

File: beacon_chain/attestation_pool.nim [hunk 2]
```diff
@@ -22,10 +23,23 @@ func init*(T: type AttestationPool, blockPool: BlockPool): T =
...
+    finalized_block_state_root = default(Eth2Digest), # This is unnecessary for fork choice but may help external components
+    justified_epoch = blockPool.headState.data.data.current_justified_checkpoint.epoch,
+    finalized_epoch = blockPool.headState.data.data.finalized_checkpoint.epoch,
+    # finalized_root = blockPool.headState.data.data.finalized_checkpoint.root
+    finalized_root = blockPool.finalizedHead.blck.root
```
Comment (arnetheduck) [inline beacon_chain/attestation_pool.nim:35]: why not use the checkpoint root? hm, they should always be the same

File: beacon_chain/attestation_pool.nim (hunk 1 above)
Comment (mratsim) [reply beacon_chain/attestation_pool.nim:30]: I'm not too sure but lighthouse is using those 2 fields as extra metadata in their clients:

https://github.com/status-im/nim-beacon-chain/blob/db92c2f2549a339be60896c3907cefdb394b5e11/beacon_chain/fork_choice/fork_choice_types.nim#L106-L109

I added them, thinking I could always transform them in to `Metadata: T` or remove them later depending on how the integration went.

File: beacon_chain/attestation_pool.nim (hunk 2 above)
Comment (mratsim) [reply beacon_chain/attestation_pool.nim:35]: They should but at genesis the headState is 0x0000...0000.
I believe this is the same issue that Danny raised when they had Prysmatic and Lighthouse disagreeing on the very first block.

**Comment by mratsim** [comment]
The CI failures are strange, maybe a new stack overflow?

File: beacon_chain/attestation_pool.nim (hunk 2 above)
Comment (arnetheduck) [reply beacon_chain/attestation_pool.nim:35]: ah ok, good to put a comment about that here then

**Pull request merged**